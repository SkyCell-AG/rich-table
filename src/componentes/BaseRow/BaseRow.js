import React, {
    useCallback,
    useState,
} from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'
import noop from 'lodash/noop'
import get from 'lodash/get'
import ChevronRight from '@material-ui/icons/ChevronRight'

import getComponentProps from 'utils/getComponentProps'

import BaseCell from 'componentes/BaseCell'

import useStyles from './BaseRow.style'

const propTypes = {
    name: PropTypes.string.isRequired,
    selectedRows: PropTypes.object, // eslint-disable-line
    rowSelected: PropTypes.string,
    uniqFieldValue: PropTypes.string.isRequired,
    rowClick: PropTypes.func.isRequired,
    rowProps: PropTypes.object, // eslint-disable-line
    columns: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            Header: PropTypes.func,
            Cell: PropTypes.func,
            props: PropTypes.object, // eslint-disable-line
        }).isRequired,
    ).isRequired,
    detailPanel: PropTypes.func,
    selectedRowId: PropTypes.string,
}

const defaultProps = {
    selectedRows: {},
    detailPanel: undefined,
    rowSelected: undefined,
    selectedRowId: undefined,
    rowProps: {},
}

const BaseRow = (props) => {
    const {
        name,
        selectedRows,
        rowSelected,
        selectedRowId,
        uniqFieldValue,
        rowClick,
        rowProps,
        columns,
        detailPanel: DetailPanel,
    } = props

    const classes = useStyles()

    const [
        isHide,
        setHide,
    ] = useState(true)
    const togglePanel = useCallback(() => {
        setHide(!isHide)
    }, [isHide])

    return (
        <div
            data-test-id="richtable-row"
            className={clsx({
                [classes.selectedRow]: selectedRowId === uniqFieldValue,
            })}
            onKeyDown={DetailPanel ? togglePanel : rowClick(rowProps)}
            onClick={DetailPanel ? togglePanel : rowClick(rowProps)}
        >
            <div
                className={clsx(
                    classes.rowContent,
                    {
                        [classes.cursorPointer]: !Array.isArray(selectedRows),
                        [classes.selectedRowContent]: rowSelected,
                    },
                )}
            >
                {
                    DetailPanel && (
                        <div className={classes.iconWrapper}>
                            <ChevronRight className={clsx(
                                {
                                    [classes.isOpenIcon]: !isHide,
                                },
                            )}
                            />
                        </div>
                    )
                }
                {
                    columns.map(({
                        Cell = BaseCell,
                        id,
                        width,
                        mapCellProps = noop,
                        component: Component,
                        propsMapper,
                        props: columnProps,
                    }) => {
                        const value = get(rowProps, mapCellProps)

                        const cellProps = typeof mapCellProps === 'string' ? {
                            children: Component
                                ? (
                                    <Component
                                        name={name}
                                        value={value}
                                        mapCellProps={mapCellProps}
                                        {...getComponentProps(
                                            propsMapper,
                                            rowProps,
                                        )}
                                    />
                                )
                                : value,
                        } : mapCellProps(rowProps)

                        return (
                            <div
                                style={{
                                    width,
                                }}
                                className={classes.cell}
                                key={`cell-${id}`}
                                data-testid={`cell-${id}`}
                            >
                                <Cell
                                    id={id}
                                    {...rowProps}
                                    {...columnProps}
                                    {...cellProps}
                                />
                            </div>
                        )
                    })
                }
            </div>
            {
                DetailPanel && (
                    <div className={clsx(
                        {
                            [classes.hideDetails]: isHide,
                        },
                    )}
                    >
                        <DetailPanel
                            name={name}
                            rowProps={rowProps}
                        />
                    </div>
                )
            }
        </div>
    )
}

BaseRow.propTypes = propTypes
BaseRow.defaultProps = defaultProps

export default BaseRow
