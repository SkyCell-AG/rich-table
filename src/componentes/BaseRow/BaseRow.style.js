import {
    makeStyles,
} from '@material-ui/core'

const useStyles = makeStyles((theme) => {
    return {
        selectedRow: {
            background: `${theme.palette.primary[50]} !important`,
        },
        rowContent: {
            display: 'flex',
            justifyContent: 'space-around',
            minHeight: '40px',
            '&:hover': {
                backgroundColor: theme.palette.primary[50],
            },
        },
        cursorPointer: {
            cursor: 'pointer',
        },
        selectedRowContent: {
            opacity: 0.7,
        },
        cell: {
            width: '100%',
            overflow: 'hidden',
            borderBottom: `1px solid ${theme.palette.secondary[100]}`,
        },
        hideDetails: {
            display: 'none',
        },
        showDetails: {
            display: 'block',
        },
        iconWrapper: {
            height: 'auto',
            borderBottom: `1px solid ${theme.palette.secondary[100]}`,
            display: 'flex',
            alignItems: 'center',
            color: theme.palette.secondary[600]
        },
        isOpenIcon: {
            transform: 'rotate(90deg)',
        }
    }
})

export default useStyles
