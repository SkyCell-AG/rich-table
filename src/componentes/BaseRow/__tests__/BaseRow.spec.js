import React from 'react'
import {
    shallow,
    mount,
} from 'enzyme'
import ChevronRight from '@material-ui/icons/ChevronRight'

import BaseRow from '../BaseRow'

jest.mock(
    '../BaseRow.style',
    () => {
        return () => {
            return {
                hideDetails: 'hideDetails',
                showDetails: 'showDetails',
            }
        }
    },
)

describe('components/BaseRow', () => {
    it('should match snapshot', () => {
        const props = {
            name: 'Note',
            columns: [
                {
                    id: 'noteSubject',
                    headerProps: {
                        children: 'Note Subject',
                    },
                    mapCellProps: 'noteSubject',
                },
                {
                    id: 'noteText',
                    headerProps: {
                        children: 'Note Text',
                    },
                    mapCellProps: 'noteText',
                },
            ],
            uniqFieldValue: '9x3C821DE9260469101EAEF912A3B3961A9E71B1D7037AF64F78DBE960',
            rowClick: jest.fn(),
            rowProps: {
                noteSubject: 'Logger Exchange',
                noteText: 'Internal: ffff477a6c5d',
                changedBy: 'alessio.conese@skycell.ch',
                validDateTimestamp: '28.12.2020 10:35:56 +0100',
                id: '1',
            },
        }

        const wrapper = shallow(<BaseRow {...props} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('with detailPanel', () => {
        const props = {
            name: 'Note',
            columns: [
                {
                    id: 'noteSubject',
                    headerProps: {
                        children: 'Note Subject',
                    },
                    mapCellProps: 'noteSubject',
                },
                {
                    id: 'noteText',
                    headerProps: {
                        children: 'Note Text',
                    },
                    mapCellProps: 'noteText',
                },
            ],
            uniqFieldValue: '9x3C821DE9260469101EAEF912A3B3961A9E71B1D7037AF64F78DBE960',
            rowClick: jest.fn(),
            rowProps: {
                noteSubject: 'Logger Exchange',
                noteText: 'Internal: ffff477a6c5d',
                changedBy: 'alessio.conese@skycell.ch',
                validDateTimestamp: '28.12.2020 10:35:56 +0100',
            },
            detailPanel: () => {
                return <div className="detailPanel">detailPanel</div>
            },
        }

        const wrapper = mount(<BaseRow {...props} />)
        const iconChevron = wrapper.find(ChevronRight)
        // const row = wrapper.find

        expect(iconChevron.length).toBe(1)
        expect(wrapper.find({
            className: 'detailPanel',
        }).length).toBe(0)

        iconChevron.simulate('click')

        expect(wrapper.find({
            className: 'detailPanel',
        }).length).toBe(1)
    })
})
