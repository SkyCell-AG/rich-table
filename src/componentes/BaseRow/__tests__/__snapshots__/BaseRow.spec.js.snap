// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BaseRow should match snapshot 1`] = `
<div
  className="cursorPointer"
  data-testid="richtable-row-9x3C821DE9260469101EAEF912A3B3961A9E71B1D7037AF64F78DBE960"
>
  <div
    className="cursorPointer undefined"
    onClick={[MockFunction]}
    onKeyDown={[MockFunction]}
  >
    <div
      data-testid="cell-noteSubject"
      key="cell-noteSubject"
      style={
        Object {
          "display": "flex",
          "flex": "0 0 20px",
          "minWidth": "",
        }
      }
    >
      <BaseCell
        changedBy="alessio.conese@skycell.ch"
        className=""
        id="1"
        noteSubject="Logger Exchange"
        noteText="Internal: ffff477a6c5d"
        validDateTimestamp="28.12.2020 10:35:56 +0100"
      >
        Logger Exchange
      </BaseCell>
    </div>
    <div
      data-testid="cell-noteText"
      key="cell-noteText"
      style={
        Object {
          "minWidth": "",
        }
      }
    >
      <BaseCell
        changedBy="alessio.conese@skycell.ch"
        className=""
        id="1"
        noteSubject="Logger Exchange"
        noteText="Internal: ffff477a6c5d"
        validDateTimestamp="28.12.2020 10:35:56 +0100"
      >
        Internal: ffff477a6c5d
      </BaseCell>
    </div>
  </div>
</div>
`;
